File: BlotoutAnalytics/Events/BOADeveloperEvents.m
//
//  BOADeveloperEvents.m
//  BlotoutAnalytics
//
//  Copyright © 2019 Blotout. All rights reserved.
//

#import "BOADeveloperEvents.h"

#import "BOAUtilities.h"
#import "BOANetworkConstants.h"
#import "BOServerDataConverter.h"
#import "BOSharedManager.h"
#import "BOEncryptionManager.h"
#import "BOASDKManifestController.h"
#import "BOFLogs.h"
#import "BOCrypt.h"

@implementation BOADeveloperEvents

+(NSDictionary*)captureEvent:(BOACaptureModel*)payload {
  @try {
    return [BOADeveloperEvents createEventObject:payload.event withType:payload.type withScreenName:payload.screenName  withEventInfo:payload.properties];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
}

+(NSDictionary*)capturePersonalEvent:(BOACaptureModel*)payload isPHI:(BOOL)phiEvent {
  @try {
    return [BOADeveloperEvents preparePersonalEvent:payload.event withScreenName:payload.screenName withEventSubcode:payload.eventSubCode withEventInfo:payload.properties isPHI:phiEvent];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
}

+(NSDictionary*)prepareServerPayload:(NSArray*)events {
  @try {
    NSMutableArray *eventData = [NSMutableArray array];
    NSDictionary *metaInfo = [BOServerDataConverter prepareMetaData];
    
    for (NSDictionary *event in events) {
      [eventData addObject:[event valueForKey:BO_EVENTS]];
    }
      
    return @{BO_META:metaInfo,BO_EVENTS:eventData};
  } @catch(NSException *exception) {
    BOFLogDebug(@"%@", exception);
  }
}

+(NSDictionary*)createEventObject:(NSString*)eventName withType:(NSString*)type withScreenName:(NSString*)screenName  withEventInfo:(NSDictionary*)eventInfo {
  @try {
   
    NSMutableDictionary *properties = [NSMutableDictionary dictionary];
    [properties addEntriesFromDictionary:eventInfo];
    NSString *screenName = (screenName != nil && screenName.length >0) ? screenName : [BOSharedManager sharedInstance].currentScreenName;
    NSMutableDictionary *event = [NSMutableDictionary dictionary];
    [event setValue:eventName forKey:BO_EVENT_NAME_MAPPING];
    [event setValue: [BOAUtilities get13DigitNumberObjTimeStamp] forKey:BO_EVENTS_TIME];
    [event setValue:[BOAUtilities getMessageIDForEvent:eventName] forKey:BO_MESSAGE_ID];
    [event setValue:[BOAUtilities getDeviceId] forKey:BO_USER_ID];
    [event setValue:screenName forKey:BO_SCREEN_NAME];
    [event setValue:[BOADeveloperEvents getScreenPayload] forKey:BO_SCREEN];
    [event setValue:type forKey:BO_TYPE];
    [event setValue:[BOSharedManager sharedInstance].sessionId forKey:BO_SESSION_ID];
    [event setValue:properties forKey:@"additionalData"];
    
    return @{BO_EVENTS:event};
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

+(NSDictionary*)getEncryptedEvent:(NSString*)publicKey withSecretKey:(NSString*)secretKey withDictionary:(NSDictionary*)event {
  @try {
    NSString *personalEncryptedData = nil;
    NSString *personalEncryptedSecretKey = nil;
    if (event == nil) {
      return nil;
    }
    
    NSData * dataToEncryptPII = [NSJSONSerialization dataWithJSONObject:event options:NSJSONWritingFragmentsAllowed error:nil];
    
    personalEncryptedData = [BOCrypt encryptDataWithoutHash:dataToEncryptPII key:secretKey iv:BO_CRYPTO_IVX];
    personalEncryptedSecretKey = [BOEncryptionManager encryptString:secretKey publicKey:publicKey];
    
    NSMutableDictionary *personalPayload = [NSMutableDictionary dictionary];
    [personalPayload setObject:personalEncryptedSecretKey forKey:BO_KEY];
    [personalPayload setObject:BO_CRYPTO_IVX forKey:BO_IV];
    [personalPayload setObject:personalEncryptedData forKey:BO_DATA];
    
    if (personalEncryptedSecretKey != nil && personalEncryptedSecretKey.length > 0 && personalEncryptedData != nil && personalEncryptedData.length > 0) {
        return personalPayload;
    }
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

+(NSDictionary*)preparePersonalEvent:(NSString*)eventName withScreenName:(NSString*)screenName withEventSubcode:(NSNumber*)eventSubcode withEventInfo:(NSDictionary*)eventInfo isPHI:(BOOL)phiEvent {
  @try {
    
    NSString *secretKey = [BOAUtilities getUUIDString];
    secretKey = [secretKey stringByReplacingOccurrencesOfString:@"-" withString:@""];
    NSDictionary *encryptedData = nil;
  
    NSString *publicKey;
    NSString *eventType;
    if (phiEvent) {
      publicKey = [BOASDKManifestController sharedInstance].phiPublickey;
      eventType = BO_PHI;
    } else {
      publicKey = [BOASDKManifestController sharedInstance].piiPublicKey;
      eventType = BO_PII;
    }
    
    encryptedData = [self getEncryptedEvent:publicKey withSecretKey:secretKey withDictionary:eventInfo];
    return [BOADeveloperEvents createEventObject:eventName withType:eventType withScreenName:screenName withEventInfo:encryptedData];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  return nil;
}

+(NSDictionary*)getScreenPayload {
  NSMutableDictionary *screenInfo = [NSMutableDictionary dictionary];
  @try {
    CGSize screenSize = [UIScreen mainScreen].bounds.size;
    [screenInfo setValue:@(screenSize.width) forKey:@"width"];
    [screenInfo setValue:@(screenSize.height) forKey:@"height"];
  } @catch (NSException *exception) {
    BOFLogDebug(@"%@:%@", BOA_DEBUG, exception);
  }
  
  return screenInfo;
}
@end

File: BlotoutAnalytics/Events/BOADeveloperEvents.swift
//
//  BOADeveloperEvents.swift
//  BlotoutAnalyticsSDK
//
//  Created by Poonam Tiwari on 22/03/22.
//

import Foundation
import UIKit
class BOADeveloperEvents: NSObject {
    
    class func captureEvent(_ payload: BOACaptureModel?) -> EventModel? {
      //  return BOADeveloperEvents.createEventObject(eventName: payload?.event ?? "", withType: payload?.type ?? "", withScreenName: payload?.screenName ?? "", withEventInfo: payload?.properties ?? [:])
        
        return BOADeveloperEvents.createEventModel(eventName: payload?.event ?? "", withType: payload?.type ?? "", withScreenName: payload?.screenName ?? "", withEventInfo: payload?.properties ?? [:])
        
    }
    
//    class func capturePersonalEvent(_ payload: BOACaptureModel?, isPHI phiEvent: Bool) -> [AnyHashable : Any]? {
//        return BOADeveloperEvents.preparePersonalEvent(payload?.event ?? "", withScreenName: payload?.screenName ?? "", withEventSubcode: payload?.eventSubCode ?? 0 , withEventInfo: payload?.properties ?? [:], isPHI: phiEvent)
//        //TODO: check event subcode default value
//    }
    
    
    class func prepareServerPayload(events: [EventModel]) -> [AnyHashable : Any]? {
            var eventData: [AnyHashable] = []
            let metaInfo = BOServerDataConverter.prepareMetaData()
            
            for event in events {
                var eventDict = [String:AnyHashable]()
               // if let value = event{//eventDict[BO_EVENTS]{
                   // eventData.insert(event, at: 0)
                eventDict =  prepareEventDict(fromEventModel: event)
                eventData.append(eventDict)
                //insert(eventDict, at: 0)
               // }
            }
            if metaInfo != nil && eventData != nil {
                return [
                    BO_META: metaInfo,
                    BO_EVENTS: eventData
                ]
            } else {
                return [:]
            }
    }
    
    class func prepareEventDict(fromEventModel:EventModel)->[String:AnyHashable]
    {
        /*var evn:String?
         var evt:NSNumber?
         var mid : String?
         var scrn : String?
         var session_id : String?
         var type:String?
         var userid:String?
         var additionalData : [AnyHashable : Any]?
         var screen : [AnyHashable : Any]?*/
        var eventDict:Dictionary = [String:AnyHashable]()
        eventDict["evn"] = fromEventModel.evn
        eventDict["evt"] = fromEventModel.evt
        eventDict["mid"] = fromEventModel.mid
        eventDict["scrn"] = fromEventModel.scrn
        eventDict["session_id"] = fromEventModel.session_id
        eventDict["type"] = fromEventModel.type
        eventDict["userid"] = fromEventModel.userid
        eventDict["additionalData"] = fromEventModel.additionalData as! AnyHashable
        eventDict["screen"] = fromEventModel.screen as! AnyHashable
        
        return eventDict
    }
    
    
    
    class func createEventObject( eventName: String, withType type: String, withScreenName screenName: String, withEventInfo eventInfo: [AnyHashable : Any]) -> [AnyHashable : Any]? {

            var properties: [AnyHashable : Any] = [:]
            for (k, v) in eventInfo { properties[k] = v }
            
            let screenName = (screenName.count > 0) ? screenName : BOSharedManager.sharedInstance.currentScreenName
            var event: [AnyHashable : Any] = [:]
            event[BO_EVENT_NAME_MAPPING] = eventName
            event[BO_EVENTS_TIME] = BOAUtilities.get13DigitNumberObjTimeStamp()
            event[BO_MESSAGE_ID] = BOAUtilities.getMessageID(forEvent: eventName)
            
            event[BO_USER_ID] = BOAUtilities.getDeviceId()
            event[BO_SCREEN_NAME] = screenName
            event[BO_SCREEN] = BOADeveloperEvents.getScreenPayload()
            event[BO_TYPE] = type
            if type == "system" {
                properties[BO_PATH] = screenName
                
              //  for var element in properties { element.BO_PATH = screenName }
            }
            event[BO_SESSION_ID] = BOSharedManager.sharedInstance.sessionId
           // for var element in event { element.BO_SESSION_ID = BOSharedManager.sharedInstance.sessionId }
            event["additionalData"] = properties
            return [
                BO_EVENTS: event
            ]

    }
    
    
    class func createEventModel( eventName: String, withType type: String, withScreenName screenName: String, withEventInfo eventInfo: [AnyHashable : AnyHashable]) -> EventModel
    {
        var properties: [AnyHashable : AnyHashable] = [:]
        for (k, v) in eventInfo { properties[k] = v }
        let screenName = (screenName.count > 0) ? screenName : BOSharedManager.sharedInstance.currentScreenName
       
        let eventTime = BOAUtilities.get13DigitNumberObjTimeStamp()
        let messageID = BOAUtilities.getMessageID(forEvent: eventName)
        
        let userID = BOAUtilities.getDeviceId()
        let session_id = BOSharedManager.sharedInstance.sessionId
        let screen = BOADeveloperEvents.getScreenPayload()
        if type == "system" {
            properties[BO_PATH] = screenName
        }
        
        var event:EventModel = EventModel(evn: eventName, evt: eventTime, mid: messageID, scrn: screenName, session_id: session_id, type: type, userid: userID, additionalData: properties, screen: screen)
        return event
    }
    
    class func getScreenPayload() -> [AnyHashable : AnyHashable]? {
        var screenInfo: [AnyHashable : AnyHashable] = [:]
            let screenSize = UIScreen.main.bounds.size
            screenInfo["width"] = NSNumber(value: Float(screenSize.width))
            screenInfo["height"] = NSNumber(value: Float(screenSize.height))
        
        return screenInfo
    }
    
}
